compiler/tests/test_identify_token.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright: Copyright nsoul97 (Soulounias Nikolaos) 2019.
+ * License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Authors:   nsoul97 (Soulounias Nikolaos)
+ */
+
+#include "../identify_token.h"
+#include "../token_name.h"
+#include "../tokens.h"
+#include <cassert>
+#include <cstdio>
+
+extern Token token;
+
+/*
+If the global variable token has its kind member set to the expected_enum, the
+function prints a success message. Otherwise, the program execution stops due to
+the assertion.
+*/
+static void test(const char *input, TOK expected_enum) {
+  identify_token(input);
+  assert(token.kind == expected_enum);
+  if (expected_enum != TOK::UNDEFINED)
+    printf("SUCCESS: function identify_token(\"%s\") sets token's kind member "
+           "to %s token.\n",
+           input, token_name(expected_enum));
+  else
+    printf(
+        "SUCCESS: function identify_token(\"%s\") sets token's kind member to "
+        "UNDEFINED token.\n\t \"%s\" starts with an integer literal, an "
+        "identifier, a blank character or a non-DIL string.\n",
+        input, input);
```

Comment:
No need for all that ;)

Reply:
Well at least issue #4 was resolved :p
I actually found the success messages helpful when a test cases went wrong, since I knew the previous successful case. On the contrary, the assert message wasn't really enlightening.

Reply:
Yes, because a better assert would show you the call stack so you could see the call path, but well... I get the point. You could implement this function as a macro to solve this issue. But let's stay in this and maybe I'll think of a better solution. Still, implementing 2 issues in the same PR is bad practice, so either implement issue #4 first or don't use it here.

Reply:
Perhaps we could create our own `void assert(bool)` function that uses the functions
```
int backtrace(void **buffer, int size);
char **backtrace_symbols(void *const *buffer, int size);
```
if the boolean argument is false to print the the call stack information before `assert(0);`.
If the boolean argument is true nothing happens.
A disadvantage of this solution is that it is OS-specific.

Reply:
Ok, another kind of long answer. If you implement `my_assert` as a function, then you lose a very important feature - the line number. Note that `assert()` is _not_ a function but a macro. In the macro definition, there is something like `printf("%d", __LINE__);`. This `__LINE__` directive tells the compiler to insert there the line in which you write this `printf`. Because the `assert()` is a define and it gets replaced wherever you call it, you can get the actual line, which is different for every "call" (because it has been replaced).. With a function, the line will always be constant because you always go the same place in source code to execute the function.
So, in your case, you could implement `test` as a macro and basically do achieve what you would like - see what test-case failed based on the line. :)

Other than that, yes implementing backtrace printing is kind of hacky, we probably don't want to that.